Archivos para revision del tp nº1

Ivexed opened the pull request:
Version para revisar, en esta instancia ya se tiene la base para seguir escalando el programa.

Ivexed committed:
Add files via upload

scaprile commented on README.md:
```diff
@@ -1,2 +1,22 @@
+- Placa NUCLEO - F401RE.
+- 2 LEDs uno para indicar el led de sistema y el otro para simular la lampara.
+- 4 botones/pulsadores.
+- 1 transistor BC548.
+- 1 resistencia de 1k.
```
De puro rompe que soy, se puede saber qué hiciste con el BC548 y la R de 1K ?

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+char combination[MAX_COMBINATION_SIZE + 1]; // Añadimos un espacio adicional para el caracter nulo
+char input_combination[MAX_COMBINATION_SIZE + 1]; // Añadimos un espacio adicional para el caracter nulo
+static bool locked = OFF;
+float LDR_value;
```
Y por qué son globales ?

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+//=============== Global init ======================
+void serial_print(const char* str) {
+    serial_port.write(str, strlen(str));
+}
```
No, no es un init...

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+}
+
+// Inicializo la UART
+void uart_init() {
```
Llamar 'serial' arriba y 'uart' acá ocasiona que luego en el código se te ocurra hacer serial_init(), uart_write() y por suerte no hay más combinaciones posibles. Deberías mantener consistencia.

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+
+// Inicializo los botones
+void button_init() {
+    button1.mode(PullDown);
+    button2.mode(PullDown);
+    button3.mode(PullDown);
+    button4.mode(PullDown);
+    enterButton.mode(PullDown);
```
Pulldown ? Cómo están conectados los pulsadores ?

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+    lamp = OFF;
+    systemLed = ON;
+    buzzer = ON;
+    thread_sleep_for(30);
```
no me gusta, no es necesario, por qué demorar en el inicio ?

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+
+bool readButton(DigitalIn &button) {
+    if (button.read() == 1) {
+        thread_sleep_for(50); // Esperar 50 ms para antirebote
```
Bloqueante. @alutenberg : your call

alutenberg replied:
Sí, estoy de acuerdo con @scaprile. Evitar por favor estos bloqueos, tal como vimos en clase.

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+}
+
+// Verifica si algún botón ha sido presionado
+bool anyButtonPressed() {
+    return readButton(button1) || readButton(button2) || readButton(button3) || readButton(button4);
+}
```
O sea que si se presionan los 4 botones esto demora 200ms en ejecutarse...
Ya sé que "se espera que esto no pase", pero por qué no hacer las cosas bien ? Si no cuesta nada ?

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+    }
+}
+
+void ldrTask() {
+    float ldrValue = LDR.read(); // Leer el valor analógico del LDR (0.0 a 1.0)
+    if (ldrValue > LDR_LEVEL) {
+        lamp = OFF;
+        serial_print("Es de día.\n");
+    } else {
+        lamp = ON;
+        serial_print("Es de noche.\n");
+    }
+}
```
Y durante el crepúsculo tenés una colección de mensajes uno atrás del noche diciendo que es de día, después de noche, después de día, y así hasta el hartazgo; con las consecuencias sobre la operación del sistema, que en este caso no es ninguna porque no hacés nada con eso... solamente la lámpara va a empezar a cambiar su luminosidad o parpadear, dependiendo de cómo responda la LDR (que si llega a ver la luz de la lámpara te empieza a oscilar todo, pero eso es otro tema); lo más rápido sería lo que te permita la UART para imprimir los mensajes y el resto de los delays que tenés por ahí

scaprile replied:
Y viendo lo de arriba, la lámpara parpadea cada 1s si el sistema está ON y 300ms si está OFF.

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+
+void sysResponse() {
+    if (systemState == ON) {
+        thread_sleep_for(300);
```
Estás muy acostumbrado a vivir en la multitarea.
Estás bloqueando por 300ms

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,213 @@
+        //systemLed = OFF;
+        serial_print("¡Bienvenido!\n");
+        systemLed = !systemLed;
+        thread_sleep_for(1000);
```
y acá por 1 segundo ? Para qué ?
Ya me perdí, pero esto no se ejecuta dentro del loop principal ? Cada vez que el sistema está ON tenés un mensaje por segundo y durante ese segundo no se puede hacer nada, ni presionar un botón, porque no lo ves ?

scaprile reviewed (changes requested):
Muy convoluted, demasiadas cosas para algo tan simple, bloqueos innecesarios.

Ivexed committed:
TP1 - Update

Se corrigieron los errores mencionados en el anterior pullRequest.
Aqui se trato de manejar menos delays, de tener en cuenta los while() bloqueantes que estaban demás y se intentar de no tener demasiadas cosas en el puerto serial.

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,222 @@
+    out_init();
+
+    while (true) {
+        passSet();
+        comparedKey();
```
Ambas funciones se ejecutan siempre, ambas operan sobre el buzzer

scaprile commented on main.cpp:
```diff
@@ -0,0 +1,222 @@
+    }
+}
+
+void ldrTask() {
+    float ldrValue = LDR.read(); // Leer el valor analógico del LDR (0.0 a 1.0)
+    if (ldrValue > LDR_LEVEL) {
+        lamp = OFF;
+    } else {
+        lamp = ON;
+    }
+}
```
Esto sigue sin resolverse, y el problema no eran los prints.
Investigá qué es histéresis, comparador, ruido.

Ivexed committed:
TP1 - RevisionRequest-3

-Se agrego una funcion que solo opere sobre el buzzer.
-Se agrego una region de histeresis para el encendido y apagado del LDR
